graphql/http.go

```diff
@@ -46,10 +90,106 @@ type httpResponse struct {
 	Errors interface{} `json:"errors"`
 }
 
+// newRequest creates activeRequest object. Method grants that request will be
+// served
+func (h *httpHandler) newRequest() (uuid.UUID, error) {
+	connID := uuid.NewV4()
+	predictedAt := time.Now().Add(h.predictedDuration)
+	h.requestsBucket = append(h.requestsBucket, activeRequest{
+		id:          connID,
+		bucketLevel: len(h.requestsBucket) + 1,
+		startedAt:   time.Now(),
+		predictedAt: predictedAt,
+	})
+	return connID, nil
+}
+
+// serveRequest decides put request into work or not. If amount of simultaneous
+// requests is more than allowed (`currentMaxRequestsLevel` level), method will
+// wait for up to `httpHandler.waitTime` and repeat attempt to start request. If
+// request started: returns UUID of request, otherwise returns empty UUID and
+// error
+func (h *httpHandler) serveRequest(isInitial bool) (uuid.UUID, error) {
+	if len(h.requestsBucket) < h.currentMaxRequestsLevel {
+		return h.newRequest()
+	}
+	if isInitial && h.predictedDuration <= h.waitTime {
+		time.Sleep(h.predictedDuration)
+		return h.serveRequest(false)
+	}
+	return uuid.Nil, NewClientError("limit is reached, please try again later")
+}
+
+// endRequest finishes request and removes it from the list of simultaneous
+// requests. Method adjustes `predictedDuration` value based on `endState` code.
+func (h *httpHandler) endRequest(connID uuid.UUID, endState endRequestState) {
+	if connID == uuid.Nil {
+		return
+	}
+	var connection activeRequest
+	for i, conn := range h.requestsBucket {
+		if conn.id == connID {
+			connection = conn
+			h.requestsBucket = append(h.requestsBucket[:i], h.requestsBucket[i+1:]...)
+			break
+		}
+	}
+	// we do not want update prediction if endState == endRequestStateError or
+	// somehow connection was not found
+	if connection.id == uuid.Nil || endState == endRequestStateError {
+		return
+	}
+	// if connection was timeouted, we would like to decrease
+	// currentMaxRequestsLevel or set it to level of simultaneous requests,
+	// which was when request has been accepted
+	if endState == endRequestStateTimedOut {
+		if connection.bucketLevel <= h.minRequests {
+			h.currentMaxRequestsLevel = h.minRequests
+		} else if connection.bucketLevel > h.currentMaxRequestsLevel {
+			h.currentMaxRequestsLevel--
```

Comment:
the level of requests bucket goes down to the level, that was when "timed out" request was accepted. In case when the volume of request bucket was already decreased during processing this one, I decrease it one more by 1